File: internal/actions/vars.go
package actions

import (
	"strings"

	"github.com/openforms/openforms/internal/config"
	"github.com/openforms/openforms/internal/definition"
	"github.com/openforms/openforms/internal/submissions"
)

// TemplateVars builds the variables available to email templates (spec §6.9).
func TemplateVars(cfg config.Config, form definition.Form, sub submissions.Submission, transition *definition.Transition) map[string]any {
	base := strings.TrimRight(cfg.BaseURL, "/")
	tr := map[string]any{}
	if transition != nil {
		tr["key"] = transition.Key
		tr["label"] = transition.Label
	}
	return map[string]any{
		"baseUrl":    base,
		"form":       map[string]any{"slug": form.Slug, "title": form.Title},
		"transition": tr,
		"submission": map[string]any{
			"id":         sub.ID.String(),
			"state":      sub.State,
			"stateLabel": sub.StateLabel,
			"data":       nonNil(sub.Data),
			"fields":     nonNil(sub.Fields),
			"url":        base + "/admin/submissions/" + sub.ID.String(),
		},
	}
}

func nonNil(m map[string]any) map[string]any {
	if m == nil {
		return map[string]any{}
	}
	return m
}

File: src/openforms/server/actions/vars.py
"""Variables available to action templates (spec §6.9)."""

from __future__ import annotations

from typing import Any

from ...definition import Form, Transition
from ...settings import Settings
from ..models.submissions import Submission


def template_vars(settings: Settings, form: Form, sub: Submission, transition: Transition | None) -> dict[str, Any]:
    base = settings.base_url.rstrip("/")
    tr: dict[str, Any] = {}
    if transition is not None:
        tr = {"key": transition.key, "label": transition.label}
    return {
        "baseUrl": base,
        "form": {"slug": form.slug, "title": form.title},
        "transition": tr,
        "submission": {
            "id": str(sub.id),
            "state": sub.state,
            "stateLabel": sub.state_label,
            "data": sub.data or {},
            "fields": sub.fields or {},
            "url": f"{base}/admin/submissions/{sub.id}",
        },
    }
